src/ReachSets/discretize.jl

```diff
@@ -1,310 +1,656 @@
+and ``V`` are known, these values might be hard to compute for some norms. See
+`Notes` below for details.
+
+Let ``Ω₀`` be the set defined as:
+```math
+Ω₀ = ConvexHull(\\mathcal{X}_0, e^{δA}\\mathcal{X}_0 ⊕ δU ⊕ αB_p)
+```
+where ``α = (e^{δ ‖A‖} - 1 - δ‖A‖)*(R_{\\mathcal{X}_0} + R_{U} / ‖A‖)`` and
+``B_p`` denotes the unit ball for the considered ``p``-norm.
+
+It is proved in [Lemma 1, 1] that the set of states reachable by ``S`` in the time
+interval ``[0, δ]``, that we denote ``R_{[0,δ]}(\\mathcal{X}_0)`` here,
+is included in ``Ω₀``:
+
+```math
+R_{[0,δ]}(\\mathcal{X}_0) ⊆ Ω₀.
+```
+
+Moreover, if ``d_H(A, B)`` denotes the Hausdorff distance between the sets ``A``
+and ``B`` in ``\\mathbb{R}^n``, then
+
+```math
+d_H(Ω₀, R_{[0,δ]}(\\mathcal{X}_0)) ≤ \\frac{1}{4}(e^{δ ‖A‖} - 1) D_{\\mathcal{X}_0} + 2α.
+```
+Hence, the approximation error can be made arbitrarily small by choosing ``δ``
+small enough.
+
+### Notes
+
+In this implementation, the infinity norm is used by default. Other usual norms
+are ``p=1`` and ``p=2``. However, note that not all norms are supported; see the
+documentation of `?norm` in `LazySets` for the supported norms.
+
+See also [`discretize_interpolation`](@ref) for an alternative algorithm that
+uses less conservative bounds.
+
+[1] Le Guernic, C., & Girard, A., 2010, *Reachability analysis of linear systems
+using support functions. Nonlinear Analysis: Hybrid Systems, 4(2), 250-262.*
+"""
+function discretize_firstorder(𝑆::InitialValueProblem,
+                                δ::Float64;
+                                p::Float64=Inf,
+                                exp_method::String="base")
+
+    # unwrap coefficient matrix and initial states
+    A, X0 = 𝑆.s.A, 𝑆.x0 
+
+    # system size; A is assumed square
+    n = size(A, 1)
+
+    Anorm = norm(Matrix(A), p)
+    κ = exp(δ*Anorm) - 1.0 - δ*Anorm
+    RX0 = norm(X0, p)
+
+    # compute exp(A*δ)
+    ϕ = exp_Aδ(A, δ, exp_method=exp_method)
+
+    if inputdim(𝑆) == 0
+        α = κ * RX0
+        □α = Ballp(p, zeros(n), α)
+        Ω0 = ConvexHull(X0, ϕ * X0 ⊕ □α)
+        return IVP(LDS(ϕ), Ω0)
+
+    elseif isaffine(𝑆)
+        Uset = inputset(𝑆)
         if Uset isa ConstantInput
             U = next_set(Uset)
-            RU = norm(U, Inf)
-            α = (exp(δ*Anorm) - 1. - δ*Anorm)*(RX0 + RU/Anorm)
-            β = (exp(δ*Anorm) - 1. - δ*Anorm)*RU/Anorm
-            Ω0 = CH(X0, ϕ * X0 + δ * U + Ball2(zeros(size(ϕ, 1)), α))
-            discr_U =  δ * U + Ball2(zeros(size(ϕ, 1)), β)
-            return DiscreteSystem(ϕ, Ω0, discr_U)
+            RU = norm(U, p)
+
+            # bloating coefficients
+            α = κ*(RX0 + RU/Anorm)
+            β = κ*RU/Anorm
+
+            # transformation of the initial states
+            □α = Ballp(p, zeros(n), α)
+            Ω0 = ConvexHull(X0, ϕ*X0 ⊕ δ*U ⊕ □α)
+
+            # transformation of the inputs
+            □β = Ballp(p, zeros(n), β)
+            Ud = map(u -> δ*u ⊕ □β, Uset)
```

Comment:
Why is it not just `δ*U ⊕ □β` as in the `VaryingInput` case? Will this iteration even terminate?

Reply:
The `map` function is useful to map inputs to inputs, see 

- https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L77
- https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L199

It is just a more readable version of `ConstantInput(δ*next_set(Uset) ⊕ □β)`.

I would have preferred to be able to use `map` also in the `VaryingInput` case and maybe even merge it with the constant input case, but the `□β` changes inside the loop so i don't know how to do it. I created this issue https://github.com/JuliaReach/Reachability.jl/issues/509 in case we have an idea later.

Reply:
To be fair, it would be `ConstantInput(δ*U ⊕ □β)`. I would argue that this is more readable.

Reply:
👍  I have changed it to `ConstantInput(δ*U ⊕ □β)`.